invoices/utils.py

```diff
@@ -173,10 +187,36 @@ def match_invoice(self, data):
             return invoice
         else:
             logger.info("Creating a new invoice: %s", invoice_key)
-            invoice, _ = Invoice.objects.update_or_create(year=data["date"].year, month=data["date"].month, client=data["client"], project=data["project"], defaults={"tags": data["project_tags"]})
+            invoice, _ = Invoice.objects.update_or_create(
+                year=data["date"].year,
+                month=data["date"].month,
+                client=data["client"],
+                project=data["project"],
+                defaults={"tags": data["project_tags"]})
             self.invoices_data[invoice_key] = invoice
             return invoice
 
+    def match_weekly_report(self, data):
+        week_number = data["date"].isocalendar()[1] # week number, 1-53
+        weekly_report_key = u"%s-%s %s - %s" % (data["date"].year, week_number, data["client"], data["project"])
+        weekly_report = self.weekly_report_data.get(weekly_report_key)
+        if weekly_report is not None:
+            logger.debug("Weekly report already exists: %s", weekly_report_key)
+            if weekly_report.tags != data["project_tags"]:
+                weekly_report.tags = data["project_tags"]
+                weekly_report.save()
+            return weekly_report
+        else:
+            logger.info("Creating a new weekly report %s", weekly_report_key)
+            weekly_report, _ = WeeklyReport.objects.update_or_create(
```

Comment:
If you are always creating, don't use `update_or_create`, use `weekly_report = WeeklyReport(...)\ weekly_report.save()` instead.

Reply:
This works similar to "match_invoice". Should it be changed too?

Reply:
I think yes - there is something wrong if the previous step did not find the report (/invoice), and it is created in this one. Also, a tiny detail in that `if` statement - there is no need to add `is not None`, writing just `if weekly_report` is a typical python idiom for checking it.